st3/lsp_utils/v1.py

```diff
@@ -0,0 +1,120 @@
+import shutil
+import sublime
+import weakref
+import os
+import subprocess
+from sublime_lib import ResourcePath
+from LSP.plugin import AbstractPlugin
+from LSP.plugin import ClientConfig
+from LSP.plugin import Response
+from LSP.plugin import Session
+from LSP.plugin import WorkspaceFolder
+from LSP.plugin.core.rpc import method2attr  # TODO: Remove this import
+from LSP.plugin.core.typing import List, Optional, Dict, Callable, Any
+
+
+def is_node_installed():
+    return shutil.which('node') is not None
+
+
+class ApiWrapper:
+    def __init__(self, plugin: AbstractPlugin):
+        self.__plugin = plugin
+
+    def on_notification(self, method: str, handler: Callable) -> None:
+        setattr(self.__plugin, method2attr(method), handler)
+
+    def on_request(self, method: str, handler: Callable) -> None:
+
+        def send_response(request_id, result):
+            session = self.__plugin.weaksession()
+            if session:
+                session.send_response(Response(request_id, result))
+
+        def on_response(this, params, request_id):
+            handler(params, lambda result: send_response(request_id, result))
+
+        setattr(self.__plugin, method2attr(method), on_response)
+
+
+class NpmClientHandler(AbstractPlugin):
+    # To be overridden by subclass.
+    package_name = None  # type: Optional[str]
+    server_directory = None  # type: Optional[str]
+    server_binary_path = None  # type: Optional[str]
+    # Internal
+    _package_cache_path = None  # type: Optional[str]
+    _cache_server_path = None  # type: Optional[str]
+
+    @classmethod
+    def setup(cls) -> None:
+        assert cls.package_name
+        assert cls.server_directory
+        assert cls.server_binary_path
+        if cls._package_cache_path:
+            return
+        cls._package_cache_path = os.path.join(sublime.cache_path(), cls.package_name)
+        cls._cache_server_path = os.path.join(cls._package_cache_path, cls.server_directory)
+
+    @classmethod
+    def cleanup(cls) -> None:
+        cls._package_cache_path = None
+        cls._cache_server_path = None
+
+    @classmethod
+    def name(cls) -> str:
+        assert cls.package_name
+        return cls.package_name[len('LSP-'):]
+
+    @classmethod
+    def configuration(cls) -> sublime.Settings:
+        if not cls._package_cache_path:
+            # https://github.com/sublimelsp/LSP/issues/899
+            cls.setup()
+        settings = super().configuration()
+        path = os.path.join(sublime.cache_path(), cls.package_name, cls.server_binary_path)
+        settings.set('command', ['node', path, '--stdio'])
```

Comment:
The `--stdio` part can come from `cls.get_binary_arguments()`. `LSP-bash` needs different value.

Reply:
Fixed.